### DISCUSSION TITLE
Avoid closing strategy on entry candle

### DISCUSSION THREAD

**Discussion by abeerzee** [kernc/backtesting.py#907] [Q&A]
It happens on rare cases, that in my strategy exit condition is also met on same bar on which I have taken entry. I was looking for a way to tell the system to not close a position on the bar it takes entry on, it should at least wait till closing of subsequent bar to decide if position need be closed or not? For example 
**Scenario:** In my strategy, I short when previous candle close go below 50EMA and cover when price close above 50EMA. It can happen previous candle happened below 50EMA. So I create a short position on opening of current candle, but with strong momentum price breaches 50EMA and close above it in same candle I had shorted. 

**What is happening:** Short is closed on opening of subsequent candle since exit condition was met in previous candle.

**What I want**:  Exit condition should only be checked on subsequent candle and not the candle on which entry was made
 
Can anyone advice how can I code above condition. Basically I need to tweak exit condition so it does not evaluate bar on which entry was made. Only from next candle onward it should do that.

**Comment by kernc** [comment]
Can you show some code?

**Reply by abeerzee** [reply]
Sure, here is some code I'm using
`
class SmaStrat(Strategy):
    trade_taken = 0
  
  def init(self):
        # Precompute the two moving averages
        self.hsma1 = self.I(talib.SMA, self.data.Close, self.n1)
        self.hsma2 = self.I(talib.SMA, self.data.Close, self.n2)
        self.dsma1 = resample_apply('D', talib.SMA, self.data.Close, self.d_n1)
        self.dsma2 = resample_apply('D', talib.SMA, self.data.Close, self.d_n2)        
    

    def next(self):
        if(self.trade_taken > 0): #pd.isna()
            last_trade = self.closed_trades[-1]
        price = self.data.Close[-1]

        if (self.dsma1 < self.dsma2  and self.data.Close[-1] < min(self.hsma1[-1], self.hsma2[-1], self.data.Close[-2])  and not self.position.is_short):
            if(self.trade_taken==0):
                self.sell(tp=0.965*price, sl=1.005*price,size=25)
                long_entry_bar = len(self.data.index)
            elif(len(self.data.index) - last_trade.exit_bar > 1): #  To introduce one bar lag after previous trade close
                self.sell(tp=0.965*price, sl=1.005*price,size=25)
                long_entry_bar = len(self.data.index)
                
        #Short Cover
        if (((self.dsma1[-1] > self.dsma2[-1]) or (self.data.Close[-1] < max(self.hsma1[-1], self.hsma2[-1]))) and self.position.is_long ): 
            self.position.close()
            self.trade_taken=self.trade_taken+1 
`
See the highlighted trade. I want to introduce another check i.e. if we take a trade, the exit condition should be evaluated on next candle and not on same candle (barring tp/sl)
![image](https://user-images.githubusercontent.com/99879458/218177424-347c0939-cad2-43b4-b6d0-9fb784010c8b.png)
